test/base-io-engine.cpp

```diff
@@ -0,0 +1,146 @@
+		BOOST_CHECK_EQUAL(called, 0);
+
+		timer.expires_from_now(boost::posix_time::seconds(2));
+		timer.async_wait(yc);
+		BOOST_CHECK_EQUAL(called, 0);
+
+		timer.expires_from_now(boost::posix_time::seconds(2));
+		timer.async_wait(yc);
+	});
+
+	io.run();
+	BOOST_CHECK_EQUAL(called, 1);
+}
+
+BOOST_AUTO_TEST_CASE(timeout_cancelled)
+{
+	boost::asio::io_context io;
+	boost::asio::io_context::strand strand (io);
+	int called = 0;
+
+	boost::asio::spawn(strand, [&](boost::asio::yield_context yc) {
+		boost::asio::deadline_timer timer (io);
+		Timeout::Ptr timeout = new Timeout(io, strand, boost::posix_time::seconds(3), [&called] { ++called; });
+
+		timer.expires_from_now(boost::posix_time::seconds(2));
+		timer.async_wait(yc);
+
+		timeout->Cancel();
+		BOOST_CHECK_EQUAL(called, 0);
+
+		timer.expires_from_now(boost::posix_time::seconds(2));
+		timer.async_wait(yc);
+	});
+
+	io.run();
+	BOOST_CHECK_EQUAL(called, 0);
+}
+
+BOOST_AUTO_TEST_CASE(timeout_scope)
+{
+	boost::asio::io_context io;
+	boost::asio::io_context::strand strand (io);
+	int called = 0;
+
+	boost::asio::spawn(strand, [&](boost::asio::yield_context yc) {
+		boost::asio::deadline_timer timer (io);
+
+		{
+			Timeout::Ptr timeout = new Timeout(io, strand, boost::posix_time::seconds(3), [&called] { ++called; });
+
+			timer.expires_from_now(boost::posix_time::seconds(2));
+			timer.async_wait(yc);
+		}
+
+		BOOST_CHECK_EQUAL(called, 0);
+
+		timer.expires_from_now(boost::posix_time::seconds(2));
+		timer.async_wait(yc);
+	});
+
+	io.run();
+	BOOST_CHECK_EQUAL(called, 0);
+}
+
+BOOST_AUTO_TEST_CASE(timeout_due_cancelled)
+{
+	boost::asio::io_context io;
+	boost::asio::io_context::strand strand (io);
+	int called = 0;
+
+	boost::asio::spawn(strand, [&](boost::asio::yield_context yc) {
+		boost::asio::deadline_timer timer (io);
+		Timeout::Ptr timeout = new Timeout(io, strand, boost::posix_time::seconds(3), [&called] { ++called; });
+
+		// Give the timeout enough time to become due while blocking its strand to prevent it from actually running...
+		Utility::Sleep(4);
+
+		BOOST_CHECK_EQUAL(called, 0);
+
+		// ... so that this shall still work:
+		timeout->Cancel();
+
+		BOOST_CHECK_EQUAL(called, 0);
+
+		timer.expires_from_now(boost::posix_time::seconds(1));
+		timer.async_wait(yc);
+	});
+
+	io.run();
+	BOOST_CHECK_EQUAL(called, 0);
+}
+
+BOOST_AUTO_TEST_CASE(timeout_due_scope)
+{
+	boost::asio::io_context io;
+	boost::asio::io_context::strand strand (io);
+	int called = 0;
+
+	boost::asio::spawn(strand, [&](boost::asio::yield_context yc) {
+		boost::asio::deadline_timer timer (io);
+
+		{
+			Timeout::Ptr timeout = new Timeout(io, strand, boost::posix_time::seconds(3), [&called] { ++called; });
+
+		   // Give the timeout enough time to become due while blocking its strand to prevent it from actually running...
+		   Utility::Sleep(4);
+
+		   BOOST_CHECK_EQUAL(called, 0);
+
+		   // ... so that this shall still work:
+		   timeout.reset();
```

Comment:
The indentation is broken and why do you need the `timeout.reset()` call at the end of the block? Shouldn't the explicit scope do this for you?

Reply:
Absolutely! I just need something to attach my comment to.